### PR TITLE
TOB-ticket-19

### PR DESCRIPTION
**Pull request by pkhivesara**
Attempts to fix: https://github.com/OpenArchive/OA-Trail-of-Bits/issues/19

@n8fr8 @tladesignz I have moved all OkHttpClient generation in one method within ``Utility``. This is under no form a complete PR yet. 

The piece where I would need input is ``generateOkHttpClient`` within ``Utility.kt``. I have added a ``if`` block in there which checks for ``use_tor`` and ``isOrbotInstalled``. 

@n8fr8 Please suggest possible solutions to connecting via tor/orbot.

### CODE REVIEW

**Comment by hiromipaw** [comment]
@pkhivesara this logic should already be present in the app, i.e. we are already using orbot to connect to Tor. Can you explain a little bit more where do you need help here, so that it is easier for @n8fr8 to guide you?

**Comment by pkhivesara** [comment]
> @pkhivesara this logic should already be present in the app, i.e. we are already using orbot to connect to Tor. Can you explain a little bit more where do you need help here, so that it is easier for @n8fr8 to guide you?

The audit team mentioned that the app is not respecting use_tor preferences. So something is not right with the current implementation of tor/orbot. I need to know how do we go about adding orbot support, if there is a new approach

**Comment by hiromipaw** [comment]
Thanks @pkhivesara. Are we checking 
``
if (Prefs.getUseTor() && OrbotHelper.isOrbotInstalled(context))
``
Everywhere where we should in the current code? I have seen you have moved the check while refactoring the code, if you move also the logic that we were using before is the setting respected?
I am just wondering if because we were performing the check in several places we were simply forgetting to check the condition properly and everywhere as opposed to something being wrong with orbot itself.

**Comment by pkhivesara** [comment]
> Thanks @pkhivesara. Are we checking `if (Prefs.getUseTor() && OrbotHelper.isOrbotInstalled(context))` Everywhere where we should in the current code? I have seen you have moved the check while refactoring the code, if you move also the logic that we were using before is the setting respected? I am just wondering if because we were performing the check in several places we were simply forgetting to check the condition properly and everywhere as opposed to something being wrong with orbot itself.

The issue is not that the check is being used. The issue is that despite having the check in place, the app is not able to transmit data securely over TOR as per the audit ticket.

**Comment by hiromipaw** [comment]
@pkhivesara are you checking also that orbot is running and not just that it is installed?

I think @tladesignz might know better but you could use this method to check if the app is installed and if it is to run it and also connect to tor:
https://github.com/guardianproject/NetCipher/blob/master/libnetcipher/src/info/guardianproject/netcipher/proxy/OrbotHelper.java#L300

**Comment by pkhivesara** [comment]
> @pkhivesara are you checking also that orbot is running and not just that it is installed?
> 
> I think @tladesignz might know better but you could use this method to check if the app is installed and if it is to run it and also connect to tor: https://github.com/guardianproject/NetCipher/blob/master/libnetcipher/src/info/guardianproject/netcipher/proxy/OrbotHelper.java#L300

We have the following on master.

```
        if (getUseTor() && OrbotHelper.isOrbotInstalled(context)) {
            try {
                OrbotHelper.requestStartTor(context)
                val builder = StrongOkHttpClientBuilder(context)
                builder.withBestProxy().build(object : StrongBuilder.Callback<OkHttpClient?> {
                    override fun onConnected(okHttpClient: OkHttpClient?) {
                        Log.i("NetCipherClient", "Connection to orbot established!")
                        client = okHttpClient
                    }

                    override fun onConnectionException(exc: Exception) {
                        Log.e("NetCipherClient", "onConnectionException()", exc)
                    }

                    override fun onTimeout() {
                        Log.e("NetCipherClient", "onTimeout()")
                    }

                    override fun onInvalid() {
                        Log.e("NetCipherClient", "onInvalid()")
                    }
                })
            } catch (exc: Exception) {
                Log.e("Error", "Error while initializing TOR Proxy OkHttpClient", exc)
            }
        }
        ```
        This uses ``StrongOkHttpClientBuilder`` which Nathan suggested we can get rid off. I just need to know how does one connect successfully via TOR. The current logic connects to the the Orbot app, but obviously the network is not secure despite connecting to Orbot as per the audit ticket.

**Comment by hiromipaw** [comment]
@pkhivesara I think the audit mention a few issues regarding which events do not go through tor. Those are linked in https://github.com/OpenArchive/OA-Trail-of-Bits/issues/19 in the code references.
I think in these cases we aren't enforcing the settings at all. I do not think that in this case this is an issue with orbot. So here is what we should do:
1. Ensure that we are checking what the user has set regarding using tor every time we do a request
2. If the connections needs to go through tor, ensure that if there is any error we can propagate the message to the user.

**Comment by pkhivesara** [comment]
> @pkhivesara I think the audit mention a few issues regarding which events do not go through tor. Those are linked in [OpenArchive/OA-Trail-of-Bits#19](https://github.com/OpenArchive/OA-Trail-of-Bits/issues/19) in the code references. I think in these cases we aren't enforcing the settings at all. I do not think that in this case this is an issue with orbot. So here is what we should do:
> 
>     1. Ensure that we are checking what the user has set regarding using tor every time we do a request
> 
>     2. If the connections needs to go through tor, ensure that if there is any error we can propagate the message to the user.

I have updated the codebase, though I have a feeling that this may not be complete. Let me know.

**Comment by hiromipaw** [comment]
> I have updated the codebase, though I have a feeling that this may not be complete. Let me know.

Can we at least add tests that the setting is respected? 
The code style bot added a few issues too. Can we address those as well?

**Comment by pkhivesara** [comment]
> > I have updated the codebase, though I have a feeling that this may not be complete. Let me know.
> 
> Can we at least add tests that the setting is respected? The code style bot added a few issues too. Can we address those as well?

All unit testing will be covered as part of the ticket 29. There is quite a bit of legwork involved in setting up the dependencies etc initially. I will branch off once all the other tickets are merged in and then work on unit testing, to avoid dealing with possible merge conflicts.

File: app/src/main/java/io/scal/secureshareui/login/ArchiveLoginActivity.kt [hunk 1]
```diff
@@ -54,18 +60,21 @@ class ArchiveLoginActivity : BaseActivity() {
...
-         *
-         * try {
-         * WebkitProxy.setProxy("android.app.Application", getApplicationContext(),mWebview,proxyHost,proxyPort) ;
-         * } catch (Exception e) {
-         * Log.e(TAG, "user selected \"use tor\" but an exception was thrown while setting the proxy: " + e.getLocalizedMessage());
-         * return;
-         * }
-         *
-         * } */
+        if (proxyHost != null && Prefs.getUseTor() && OrbotHelper.isOrbotInstalled(this)) {
```
[the remark this answers is not in the document]
Comment (tladesignz) [reply app/src/main/java/io/scal/secureshareui/login/ArchiveLoginActivity.kt:63]: This is still wrong. (Will comment in detail later.)

**Review by tladesignz — CHANGES_REQUESTED** [review]

**Comment by tladesignz** [comment]
@pkhivesara, I'm sorry. It seems, I didn't explain it good enough, yet.

Can I please direct your attention to the copies of these lines:

```kotlin
if (Prefs.getUseTor() && OrbotHelper.isOrbotInstalled(context))
```

This condition means, that it checks, if the user wants to use Tor, but it only honours it, **if Orbot is actually installed**.

 **Otherwise** it **ignores** the `Prefs.getUseTor()` setting as *if it was never selected in the first place*.

This is wrong and the root cause of the ToB-ticket.

You will need to rearrange the code so it does the first check after the other like this:

```kotlin
if (Prefs.getUseTor()) {
    if (!OrbotHelper.isOrbotInstalled(context)) {
        // TODO: inform the user to install Orbot.
        return
    }

    if (!OrbotHelper.isOrbotRunning(context)) {
        // TODO: Start Orbot and DON'T ALLOW uploads until Orbot runs!
        return
    }

    // All fine, continue with upload.
}
```

Please check out the link provided by @hiromipaw further up:
https://github.com/guardianproject/NetCipher/blob/master/libnetcipher/src/info/guardianproject/netcipher/proxy/OrbotHelper.java#L300
on how to start Orbot.

And while we're at it, please remove the copy of this check. **Do not duplicate this code**.

Please let me know, if you have any more questions!

**Comment by pkhivesara** [comment]
> @pkhivesara, I'm sorry. It seems, I didn't explain it good enough, yet.
> 
> Can I please direct your attention to the copies of these lines:
> 
> ```kotlin
> if (Prefs.getUseTor() && OrbotHelper.isOrbotInstalled(context))
> ```
> 
> This condition means, that it checks, if the user wants to use Tor, but it only honours it, **if Orbot is actually installed**.
> 
> **Otherwise** it **ignores** the `Prefs.getUseTor()` setting as _if it was never selected in the first place_.
> 
> This is wrong and the root cause of the ToB-ticket.
> 
> You will need to rearrange the code so it does the first check after the other like this:
> 
> ```kotlin
> if (Prefs.getUseTor()) {
>     if (!OrbotHelper.isOrbotInstalled(context)) {
>         // TODO: inform the user to install Orbot.
>         return
>     }
> 
>     if (!OrbotHelper.isOrbotRunning(context)) {
>         // TODO: Start Orbot and DON'T ALLOW uploads until Orbot runs!
>         return
>     }
> 
>     // All fine, continue with upload.
> }
> ```
> 
> Please check out the link provided by @hiromipaw further up: https://github.com/guardianproject/NetCipher/blob/master/libnetcipher/src/info/guardianproject/netcipher/proxy/OrbotHelper.java#L300 on how to start Orbot.
> 
> And while we're at it, please remove the copy of this check. **Do not duplicate this code**.
> 
> Please let me know, if you have any more questions!

Two issues here:
1. The guardian project link is incomplete. It talks about booting up Orbot, but does not show a complete example of what are the possible status code, and what needs to be handled in ``onActivityResult``. 

2. The TOB guys suggested a refactor for generating the HTTP clients. Your comment suggests that the utility method itself takes care off the upload. 

Maybe I am not saying this correctly, but my initial question is still unanswered here. 
Based on this link: https://github.com/guardianproject/NetCipher/blob/master/libnetcipher/src/info/guardianproject/netcipher/proxy/OrbotHelper.java#L300, any app that needs to route traffic securely, needs to check for Orbot and verify if it is installed AND running. 

Is that all to this??

To my untrained eye, it seems like we are still missing some core logic here. What is the difference in HTTP client generation here for secure traffic routing if we are using the same ``OkHttpClientBuilder`` for TOR and no-TOR scenario?

**Comment by tladesignz** [comment]
> Two issues here:
> 
> 1. The guardian project link is incomplete. It talks about booting up Orbot, but does not show a complete example of what are the possible status code, and what needs to be handled in `onActivityResult`.
> 2. The TOB guys suggested a refactor for generating the HTTP clients. Your comment suggests that the utility method itself takes care off the upload.
> 
> Maybe I am not saying this correctly, but my initial question is still unanswered here. Based on this link: https://github.com/guardianproject/NetCipher/blob/master/libnetcipher/src/info/guardianproject/netcipher/proxy/OrbotHelper.java#L300, any app that needs to route traffic securely, needs to check for Orbot and verify if it is installed AND running.
> 
> Is that all to this??
> 
> To my untrained eye, it seems like we are still missing some core logic here. What is the difference in HTTP client generation here for secure traffic routing if we are using the same `OkHttpClientBuilder` for TOR and no-TOR scenario?

@pkhivesara, I'm sorry, while I do have quite some experience with Android, as I actually started out my ventures into mobile devices with Android, I never came round to actually develop something from the grounds up which uses Tor or Orbot, so I'm not yet completely familiar with the Android Tor/Orbot ecosystem.

Nate will probably not going to help you with your generic questions, either, as he's pretty busy managing Guardian Project. (Busy as in: he sometimes works for 3 persons. No idea, where he takes the power for that, but he's amazing.)

What I do in that cases is, that I start reading the [general documentation](https://github.com/guardianproject/NetCipher/blob/master/README.md) and after that look at the inline docs and source code.
Don't worry, that's cool. You're paid to do researchy development here, not for being a code monkey.

The way you phrase your question also suggests, that you're not using Orbot. I suggest, you get yourself acquainted with it to understand, what it is capable of.

Additional hint: Look at `NetCipher#StrongOkHttpClientBuilder`. If it's still fresh enough, use that. If it turns out, it's outdated, please copy what it does to the latest OkHttpClient in your own code as best as possible.

**Comment by n8fr8** [comment]
Hi all. Jumping into help as I can. Some core concepts to make sure everyone is on the same page.

- To route an app through Tor on Android, you need to proxy network traffic either through either the local SOCKS proxy or HTTP proxy that are provided by Orbot. These usually run at port 9050 (Socks) and 8118 (HTTP) on localhost or 127.0.0.1.

- The NetCipher/OrbotHelper libraries are just there to make things a bit smoother for the user. If you want to simplify this work, you can just start with just focusing on enabling proxy features. This is what Tusky does (Mastodon client), and Twitter also did for many years. Some users may also enjoy the ability from working with any kind of proxy server, not just Orbot.

- The NetCipher/OrbotHelper interaction can be used to determine 1) is a legitimate signed instance of Orbot installed 2) ask it start and 3) find out what ports it is running local proxies on. That is where the NetCipher documentation can be helpful. 

You can see some example use of NetCipher integration in the F-Droid app here:
https://gitlab.com/fdroid/fdroidclient/-/blob/master/app/src/main/java/org/fdroid/fdroid/FDroidApp.java#L508
https://gitlab.com/fdroid/fdroidclient/-/blob/master/app/src/main/java/org/fdroid/fdroid/net/DownloaderFactory.java#L36

In particular, NetCipher should receive the ports from Orbot callback events, that allow it to provide you with the available proxy settings through a simple method call.

As has been said, the crux of this ticket is that the user thinks they are using Tor when Orbot is not installed / activated. That is what should be solved first, and then these subsequent implementations details can be reviewed to make sure the user of Orbot/Tor is fully functoining.

**Comment by pkhivesara** [comment]
@n8fr8 @tladesignz Thanks for the input. Nathan had suggested that we could get rid of ``NetCipher#StrongOkHttpClientBuilder`` in the refactor. To close this ticket for now, I will do the following:

1. Update the codebase to check if ``use_tor`` is enabled.
2. Check if Orbot is installed. If not, perhaps redirect the user to the app store??? (Do we want to do this?)
3. Check if Orbot is running. If not, show a error message to the user.

@hiromipaw Let me know if these steps sound like a step in the right direction to close this ticket.

**Comment by hiromipaw** [comment]
@pkhivesara yes those steps are a move in the right direction.
1. Check if the settings use_tor is enabled. If so make sure the traffic is routed through tor.
2. Check if Orbot is installed. If not, ask the user if they want to be directed to the play store.
3. Check if orbot is running. If not, show an error message to the user. You should check the orbot helper methods here. I think they have a method to take the user to orbot and eventually enable it.

**Comment by pkhivesara** [comment]
@tladesignz @hiromipaw I have moved this PR into ready to review. Let me know if I missed anything. There are couple of detekt flags that I dont necessarily agree with.

**Pull request closed without merging**